src/Middleware/HttpLogging/src/HttpLoggingOptions.cs

```diff
@@ -0,0 +1,78 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Text;
+using Microsoft.Net.Http.Headers;
+
+namespace Microsoft.AspNetCore.HttpLogging
+{
+    /// <summary>
+    /// Options for the <see cref="HttpLoggingMiddleware"/>.
+    /// </summary>
+    public sealed class HttpLoggingOptions
+    {
+        /// <summary>
+        /// Fields to log for the Request and Response. Defaults to logging request and response properties and headers.
+        /// </summary>
+        public HttpLoggingFields LoggingFields { get; set; } = HttpLoggingFields.RequestPropertiesAndHeaders | HttpLoggingFields.ResponsePropertiesAndHeaders;
+
+        /// <summary>
+        /// Request header values that are allowed to be logged.
+        /// <p>
+        /// If a request header is not present in the <see cref="RequestHeaders"/>,
+        /// the header name will be logged with a redacted value.
+        /// </p>
+        /// </summary>
+        public ISet<string> RequestHeaders => _internalRequestHeaders;
+
+        internal HashSet<string> _internalRequestHeaders = new HashSet<string>(StringComparer.OrdinalIgnoreCase)
+        {
+            HeaderNames.Accept,
+            HeaderNames.AcceptEncoding,
+            HeaderNames.AcceptLanguage,
+            HeaderNames.Allow,
+            HeaderNames.Connection,
+            HeaderNames.ContentLength,
+            HeaderNames.ContentType,
+            HeaderNames.Host,
+            HeaderNames.UserAgent
+        };
+
+        /// <summary>
+        /// Response header values that are allowed to be logged.
+        /// <p>
+        /// If a response header is not present in the <see cref="ResponseHeaders"/>,
+        /// the header name will be logged with a redacted value.
+        /// </p>
+        /// </summary>
+        public ISet<string> ResponseHeaders => _internalResponseHeaders;
+
+        internal HashSet<string> _internalResponseHeaders = new HashSet<string>(StringComparer.OrdinalIgnoreCase)
+        {
+            HeaderNames.ContentLength,
+            HeaderNames.ContentType,
+            HeaderNames.TransferEncoding
+        };
+
+        /// <summary>
+        /// Options for configuring encodings for a specific media type.
+        /// <p>
+        /// If the request or response do not match the supported media type,
+        /// the response body will not be logged.
+        /// </p>
+        /// </summary>
+        public MediaTypeOptions MediaTypeOptions { get; } = MediaTypeOptions.BuildDefaultMediaTypeOptions();
+
+        /// <summary>
+        /// Maximum request body size to log (in bytes). Defaults to 32 KB.
+        /// </summary>
+        public int RequestBodyLogLimit { get; set; } = 32 * 1024;
```

Comment:
Do we need Size in this name? We have size everywhere else in Kestrel.

```C#
long RequestBodySizeLimit
long ResponseBodySizeLimit
```

Why didn't we go with longs, we didn't learn from IIS 😄

Reply:
It's a long in Kestrel. I can only imagine the logs with 2 GB+ bodies inside 😆

Reply:
We could even make it nullable for consistency with Kestrel. Why force an upper bound of 2^63 bytes!

Reply:
Is this a serious conversation? I guess we can make it a long....

Reply:
OK we can leave it as an int and rename it to SizeLimit

Reply:
We covered this in the API review and picked LogLimit over SizeLimit. In kestrel it's an absolute size limit and exceeding it causes the request to fail. Here it's a limit on the amount that will be logged before we truncate.

You really want to support GB of logged data? We'll definitely need to switch to a chunking/streaming model then to avoid buffering the whole thing.

Reply:
Lets stick with an int (2GB is still GB) but I think Size should be in the name. It's the size of the request and response body in bytes.

Reply:
I agree with @Tratcher, unless we have a name like `RequestBodyLogSizeLimit` which seems needlessly verbose.